### PR TITLE
Dockerize the "Remote GPU" service

### PR DESCRIPTION
**Pull request by mikaelhg**
This PR needs some testing before merging, so I'm creating it as a draft.

Based on my previous contribution to AliaksandrSiarohin/first-order-model#55.

## Building

```
docker build -t avatarify .
```

## Running

```
docker run -it --rm --gpus=all -p 5557:5557 -p 5558:5558 avatarify
```

### CODE REVIEW

**Comment by mintmaker** [comment]
Works, but with 15-20 fps instead of ~30 fps (1080 Ti, Ubuntu 20.04). Probably because of network speed as nvidia-docker is just a bit behind native installations of tensorlfow (and probably pytorch, too).
A way to forward the video devices directly would be awesome.

But for remote GPU server it's great.

**Comment by mikaelhg** [comment]
@mintmaker, would you mind testing with `docker run` parameters `--shm-size=1g --ulimit memlock=-1 --ulimit stack=67108864` please?

**Comment by mintmaker** [comment]
runs, but the performance does not improve. With the first avatar it runs with ~23 fps, but after switching it goes down to around ~16 fps and goes up after that.

If possible (not an expert in Docker) everything dockerized would be super cool. With the current solution the worker has to install Miniconda/Anaconda and some modules, which is okay in the convential case, but not optimal for a local setup.

**Comment by mikaelhg** [comment]
I'm currently experimenting with 

```
docker run -it --rm --privileged \
    --env="DISPLAY" --env="QT_X11_NO_MITSHM=1" \
    --volume="/tmp/.X11-unix:/tmp/.X11-unix:rw" \
    -v /tmp/models:/root/.torch/models \
      avatarify python3 afy/cam_fomm.py \
          --config fomm/config/vox-adv-256.yaml \
          --checkpoint vox-adv-cpk.pth.tar --virt-cam 9 \
          --relative --is-client \
          --in-addr tcp://192.168.1.45:5557 --out-addr tcp://192.168.1.45:5558
```

**Comment by mintmaker** [comment]
Nice that you corrected the thing with -p, I forgot to mention I needed to change it.

I changed the ip you wrote with mine and the client and server were able to connect (the server downloaded stuff, he only does it when he is connected with client). But the /dev/video9 does not exist.
I tried to make with sudo modprobe v4l2loopback exclusive_caps=1 video_nr="9" card_label="avatarify", but it does not generate a new one. (The first one I did was sudo modprobe v4l2loopback devices=1 and it worked, but the following commands not). I will try it again after a reboot.

**Comment by mikaelhg** [comment]
You don't have to reboot, you can just `rmmod` the module, and `modprobe` it back with new options.

**Comment by mintmaker** [comment]
I thought removing the module would cause a reinstall, so thanks.

Yeah, now it works with /dev/video9. But still 25 fps.  --network=host makes it fast first (40 fps; which is weird, I don't even get it with my native install), but then after zooming in and recalibrating goes down to 25 fps.

Testing with:
`docker run -it --rm --privileged --gpus all\
    --env="DISPLAY" --env="QT_X11_NO_MITSHM=1" \
    --volume="/tmp/.X11-unix:/tmp/.X11-unix:rw" \
      avatarify python3 afy/cam_fomm.py \
          --config fomm/config/vox-adv-256.yaml \
          --checkpoint vox-adv-cpk.pth.tar --virt-cam 9 \
          --relative`
is successful. Runs with 33 fps (same as native installation).

Thank you!

And btw, what about a parameter downloading in a dir, which is mounted into docker? It always takes around half a minute to download the weights.

**Comment by mikaelhg** [comment]
See `-v /tmp/models:/root/.torch/models` in the previous comment.

**Comment by mikaelhg** [comment]
@alievk, based on testing, we'd need to add some documentation on how to use the Docker image, where should that go?

**Comment by mintmaker** [comment]
Wow, thank you very much works!

**Comment by mintmaker** [comment]
@mikaelhg What about using docker when --docker is specified in run.sh? And for installing the same.

**Comment by alievk** [comment]
> @alievk, based on testing, we'd need to add some documentation on how to use the Docker image, where should that go?

You can add Docker subsection to the Install section in the Readme.

**Comment by mintmaker** [comment]
@mikaelhg  I am currently updating the README and run.sh.

**Comment by mikaelhg** [comment]
@alievk, are you planning to add release tags to git? I'm currently pointing to specific commits in the `Dockerfile` as build parameters, which can be changed on the `docker build` command line.

**Comment by mintmaker** [comment]
I have README.md and run.sh updated for docker. Test are successful.

**Comment by mintmaker** [comment]
I am now working on support for remote gpu.
I have a few questions @alievk : Should I add the new subsections to the table of contents? Also the wiki needs then to be updated. How should we then merge my and @mikaelhg 's changes (maybe merge this pr and then merge my pr with README and run.sh?)

**Comment by mikaelhg** [comment]
Marked the review ready, let me know if you want some changes or documentation in this PR in addition to @mintmaker's work.

**Comment by mintmaker** [comment]
I updated run.sh for remote gpu and Docker. Tests are succesfull, once again.

**Comment by mintmaker** [comment]
@alievk @mikaelhg I am opening the PR with updated readme and run.sh now.

**Comment by mikaelhg** [comment]
@mintmaker, I added you as a collaborator to my fork, so you can push changes to this branch directly, and they'll be added to this PR, if you'd rather do that.

**Comment by mintmaker** [comment]
Or should I merge your fork first into mine?

**Comment by mintmaker** [comment]
@mikaelhg Okay, thanks!

**Comment by bjarthur** [comment]
i wonder if the performance would be better with singularity instead of docker

**Comment by bjarthur** [comment]
do you plan on pushing the image to the cloud?  that'd be great

**Comment by bjarthur** [comment]
https://hub.docker.com/r/bijanmmarkes/avatarify

File: README.md [hunk 1]
```diff
@@ -87,6 +89,18 @@ bash scripts/install.sh
 ```
 3. [Download network weights](#download-network-weights) and place `vox-adv-cpk.pth.tar` file in the `avatarify` directory (don't unpack it).
 
+### Docker installation
+1. Install Docker following the [Documentation](https://docs.docker.com/engine/install/). Then run this [step]{https://docs.docker.com/engine/install/linux-postinstall/#manage-docker-as-a-non-root-user} to make docker available for your user.
```
Comment (alievk) [inline README.md:93]: Please fix the markup notation for [step]()

File: README.md [hunk 2]
```diff
@@ -74,6 +74,8 @@ Download model's weights from [Dropbox](https://www.dropbox.com/s/t7h24l6wx9vret
 #### Linux
 Linux uses `v4l2loopback` to create virtual camera.
 
+### Native installation
```
Comment (alievk) [inline README.md:77]: I would leave Linux section as is and add another section called Docker on the same level, like
### Linux
...
### Docker

File: README.md [hunk 3]
```diff
@@ -158,11 +172,16 @@ The run script will create virtual camera `/dev/video9`. You can change these se
 <!--It is supposed that there is only one web cam connected to the computer at `/dev/video0`.-->
 You can use command `v4l2-ctl --list-devices` to list all devices in your system.
 
-Run:
```
Comment (alievk) [inline README.md:161]: I would leave this as is and add below a comment that one should use `--docker` option for Docker installation.

Comment (alievk) [inline README.md:161]: I would leave this as is and add below a comment that one should use `--docker` option for Docker installation.

**Comment by alievk** [comment]
> @alievk, are you planning to add release tags to git? I'm currently pointing to specific commits in the `Dockerfile` as build args, which can be changed on the `docker build` command line.

This is a good point. We can add a tag after merging this PR and change the default commit to this tag in the Dockerfile.

**Comment by alievk** [comment]
> I am now working on support for remote gpu.
> I have a few questions @alievk : Should I add the new subsections to the table of contents? Also the wiki needs then to be updated. How should we then merge my and @mikaelhg 's changes (maybe merge this pr and then merge my pr with README and run.sh?)

I can't give an access to the Wiki only, so you could create a page in your work and I'll copy-paste it to the original repo.

**Comment by alievk** [comment]
@mintmaker is there any reason not to use `--gpus` flag?

**Comment by mintmaker** [comment]
@alievk Not sure what you mean, The ```--gpus``` flag is only used with docker, so it's not important when running natively. If docker is utilized the ```--gpus all``` flag is added in the ```docker run [...]``` call, which assumes nvidia-docker is installed (it should be installed only when one owns a gpu) and adding it without nvidia-docker installed could cause some issues. I can't verify it because I have nvidia-docker already installed. Can you or @mikaelhg run it without nvidia-docker installed but with ```--gpus``` flag? If no issues are cause I can remove the flag.

**Comment by mikaelhg** [comment]
> Can you or @mikaelhg run it without nvidia-docker installed but with `--gpus` flag? If no issues are cause I can remove the flag.

```
$ docker run -it --rm --gpus=all ubuntu:20.04 bash
docker: Error response from daemon: could not select device driver "" with capabilities: [[gpu]].
```

**Comment by alievk** [comment]
I have this

``` bash
$ ./run.sh  --is-worker --no-vcam --docker
xhost:  unable to open display ""
docker: Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post http://%2Fvar%2Frun%2Fdocker.sock/v1.40/containers/create: dial unix /var/run/docker.sock: connect: permission denied.
See 'docker run --help'.
xhost:  unable to open display ""
```

Why is it trying to access a display? Is that necessary?

**Comment by mintmaker** [comment]
In order to display the preview windows it needs to acces the X Screen. As a worker this is not needed and will cause the error you just posted. I will fix this.

**Comment by alievk** [comment]
@mintmaker works perfectly!

One last thing I wanted to mention is that Avatarify is not supposed for running on CPU, so we have to run on GPU by default and stress that in the installation section (merge 1 and 2 steps). If you want to keep the `--gpus` option for some reason, I would change it to `--no-gpus`, because the default behaviour is run on a GPU. If we keep everything as is people would post issues complaining for slow performance ("nobody reads the manual").

P.S. sorry for late responds, very busy with other stuff!

**Comment by adelin-b** [comment]
I found a bug in the branch, when using docker, adding a new face, reloading with `L` and then changing face with `D` or `A` it crashes, here is the error : 
```python
[1594162784.345596] Images reloaded
Traceback (most recent call last):
  File "afy/cam_fomm.py", line 315, in <module>
    change_avatar(predictor, avatars[cur_ava])
  File "afy/cam_fomm.py", line 85, in change_avatar
    avatar_kp = predictor.get_frame_kp(new_avatar)
  File "/app/avatarify/afy/predictor_local.py", line 92, in get_frame_kp
    kp_landmarks = self.fa.get_landmarks(image)
  File "/usr/local/lib/python3.6/dist-packages/face_alignment/api.py", line 107, in get_landmarks
    return self.get_landmarks_from_image(image_or_path, detected_faces)
  File "/usr/local/lib/python3.6/dist-packages/torch/autograd/grad_mode.py", line 43, in decorate_no_grad
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/face_alignment/api.py", line 134, in get_landmarks_from_image
    if image.ndim == 2:
AttributeError: 'list' object has no attribute 'ndim'
non-network local connections being removed from access control list
```

**Comment by alievk** [comment]
Hi @adelin-b, this is not relevant to the docker branch. Could you please write about the bug in the issues?

Please also explain how to reproduce this bug and attach the avatar image if possible.

**Comment by mintmaker** [comment]
@adelin-b I fixed this issue a few commits (#221 ) before, but the Dockerfile clones the git repository one commit before my fix. This should be updated before merging, for sure.

**Comment by mikaelhg** [comment]
@alievk mentioned the possibility of tagging known good commits as release versions, that would be the way to go.

I think we've pretty much exhausted the scope of this PR, so after that change, it would probably be time for a GO/NO-GO decision.

**Comment by mintmaker** [comment]
@mikaelhg I'll write the wiki page and I think the part for docker is done then.
@alievk I'd like to recommend the docker based method in the wiki as it is easier and cleaner to setup, is that okay?

**Comment by alievk** [comment]
> @mikaelhg I'll write the wiki page and I think the part for docker is done then.
> @alievk I'd like to recommend the docker based method in the wiki as it is easier and cleaner to setup, is that okay?

that's ok

**Comment by mintmaker** [comment]
@alievk I forgot to add the installation of v4l2loopback in the README. So I'll make another script `install_docker.sh`. In your `install.sh` you clone from your fork of the original repo, does it have a reason?

**Comment by alievk** [comment]
@mintmaker I had a small fix to the upstream repository, changing the minimum version of Linux kernel https://github.com/alievk/v4l2loopback/commit/9dc107926fa3affdfad5b4b4d2b2870db1c00e4e, but looks like he already fixed that in the master branch https://github.com/umlaeute/v4l2loopback/blob/master/v4l2loopback.c#L721

Before the fix there was an error when creating a virtual camera on earlier Linux kernel versions.

**Comment by alievk** [comment]
I think you can stick to the upstream repository.

**Comment by mintmaker** [comment]
My updated wiki is here (only Remote-GPU.md is changed): https://github.com/mintmaker/avatarify_docker_wiki

**Comment by alievk** [comment]
@mintmaker excellent!

I think we are ready to merge. I can make a tag called `docker` then we have to replace the commit id with the tag in the Dockerfile. Is that ok?

**Comment by mikaelhg** [comment]
Let's go!

**Comment by alievk** [comment]
@mintmaker @mikaelhg thank you very much!

Next I'll make a tag, fix the Dockerfile, update the wiki and mention you in the readme.

**Comment by josharmour** [comment]
I'm having issues with this and willing to help troubleshoot.

**Pull request merged**